### ISSUE TITLE
document/standardize output for build status per project

### ISSUE THREAD

**Issue by marianoguerra** [ninenines/erlang.mk#387]
if the build process for all the projects in the erlang.mk library dumped the status for each project and also information about the erlang version being used I could build a nice page to display build status per project per erlang version, this would help detect breaking backwards compatibility changes and also status of support for the newest erlang versions.

**Comment by essen** [comment]
Would you like to work on the erlang.mk website also by any chance? I want something pretty simple but I'm really out of the loop with client side stuff.

**Comment by marianoguerra** [comment]
sure, do you have any ideas of what you want, sketch, drawing, page you want to base the design on?

**Comment by marianoguerra** [comment]
or maybe just content and I make a first prototype

**Comment by essen** [comment]
I'll get you that later today.

**Comment by essen** [comment]
I pushed the terrible things I made to https://github.com/ninenines/erlang.mk/tree/gh-pages

You can see the result live on http://erlang.mk/

Terrible!

Anyway this is basically what I want. Nothing big. Just prettier and with more things.

The guide is generated from Asciidoc and will need some preprocessing and styling. Tutorials will be the same, although we don't have tutorials yet (and they'll probably just be included directly in the guide anyway).

Packages can be generated from index/* and then we can add build results later on. We can probably generate a single page with all packages listed (and some JS to search/filter them), and then one page per package. Not sure about the single page though.

Overall I expect no third party dependency on CSS/JS libraries. And it should stay minimalist. CSS/JS can probably just stay embedded so as to avoid extra requests and speed up the loading faster.

Thoughts?

**Comment by marianoguerra** [comment]
Hi, this is just an initial setup, landing page:

![erlangmk-3](https://cloud.githubusercontent.com/assets/68463/10411571/c4e74778-6f6a-11e5-82ec-8881d5ba7576.png)

blog:

![erlangmk-2](https://cloud.githubusercontent.com/assets/68463/10411575/d8bf6bcc-6f6a-11e5-9183-4bdf05020eb0.png)

The theme is highly customizable

mobile version

![erlangmk-4](https://cloud.githubusercontent.com/assets/68463/10411594/3c8a7ec6-6f6b-11e5-9164-ce1b36a8b278.png)

mobile with navigation expanded:

![erlangmk-5](https://cloud.githubusercontent.com/assets/68463/10411597/421820be-6f6b-11e5-9f2b-b94d177c2305.png)

**Comment by essen** [comment]
Yeah but what happened to the colors and the logo I set? I didn't put them that way for no reasons. :-)

**Comment by essen** [comment]
Also there won't be a blog, it would require someone to maintain it and resources are better spent maintaining the tool or docs than a blog.

**Comment by marianoguerra** [comment]
blog removed, applied colors, added internal pages and published temporarily here: http://marianoguerra.org/tmp/erlangmk/

the page loads slowly because the hosting sucks :D

let me know what you think

**Comment by essen** [comment]
Yeah but I mean I want something with the logo in the middle and links around it, not the usual boring website. :-)

**Comment by marianoguerra** [comment]
![screenshot from 2015-10-10 20-01-18](https://cloud.githubusercontent.com/assets/68463/10412562/b7159d4c-6f89-11e5-964c-e1ce44bf8e1b.png)

I can try removing the navbar on the main site but on the other sections it's useful to have it.

**Comment by marianoguerra** [comment]
initial implementation of package list and status, status of course is random for now:

http://marianoguerra.org/tmp/erlangmk/packages/

![screenshot from 2015-10-10 20-47-45](https://cloud.githubusercontent.com/assets/68463/10412748/2f3618b4-6f90-11e5-8a9a-90ed835a6f90.png)

**Comment by essen** [comment]
I will ask a friend about this. Noticed you are using some kind of generator which isn't quite what I had in mind. A generator won't help for anything: we have a single landing page, guide/tutorials generated from asciidoc (just need a header and footer to put around basically), package list to be generated from the index and build results and... that's it really. The help I need is about designing that landing page and then it can be reused for everything else.

For packages, well, that's not quite enough also. There's more status than that, and more importantly, there's more OS/versions. I'm not sure everything will or should fit in one list. Also another important piece of info is the build date/time. And to go back to the design aspect, well you're back to your top navbar like every other ugly website out there. Sorry but if I wanted that I would just have used bootstrap. :-)

**Comment by essen** [comment]
And yes, I know it uses bootstrap and jquery and ... in other words all the things that I don't want. :-)

**Comment by marianoguerra** [comment]
the site generator supports using asciidoc, you need some tool to generate from asciidoc to html which is what a static site generator does... otherwise you are building a custom static site generator for your site.

regarding packages, it is the first iteration, of course it's not enough, just need to keep adding info from here, if we don't start from something it's hard to discuss in the abstract.

about bootstrap, it's not a requirement, each theme uses a different css, some may not use bootstrap, but I don't see the problem of using it, it solves all the browser quirks and adds mobile/responsive support for free.

about the header, since the themes can be customized I can remove the navbar, but when you are outside the landing page some kind of navigation to move between pages is helpful.

also jquery is not required if not used, that means we don't require it.

if you don't like the approach it's ok, but at the end we will end up creating a custom static site generator that supports asciidoc instead of using one that does :)

**Comment by essen** [comment]
The documentation is generated by a2x itself already, there's nothing to do there except provide a bit of configuration/styling. The documentation is also generating PDFs and, later on, any other format people may want (html help, epub, ...), using the same process.

I do not want to use any css or js library because they make everything a mess. In particular bootstrap requires the HTML to be built in a specific way which is a big problem. I also want to avoid any external script or css files entirely: everything in one HTML. It's fine because there should be very little anyway (and other than packages list I don't see any page need JS).

There's no need for the nav bar, the logo and links around are meant to be used in all pages.

Sure we will have a small make rule to put the files in the right place (git clone ; make docs ; cp, basically) and another one to generate the packages page (git clone ; magic > packages/index.html) but it took you longer to make this with the generator than it would to write those rules.

On the other hand what does take me a long time is making a nice design for that landing page, in particular the "logo in the center and links around" (the rest shouldn't be too difficult) and that's what I need help with.

**Comment by essen** [comment]
After a small hint from my previously mentioned friend, this is what I got: http://erlang.mk/

I have plans to make that number of package updated every time we build the package list.

Next step will be embedding the user guide though. And before that I need to take care of a few tickets and continue working on Cowboy.

**Comment by essen** [comment]
I have since pushed the guide (missing a few styles here and there, but nothing big) and am now working on the page listing all packages. I think I will make an option to switch between description and build status dynamically (making this the only JS on the whole site hehe).

**Comment by crownedgrouse** [comment]
Loic,
Geas now can give a list of possible erlang release that can compile and run a global project including deps, even from source.
```
geas:w2l(geas:compat(".", global)).
["R15","R15B","R15B01","R15B02","R15B03","R15B03-1","R16B",
"R16B01","R16B02","R16B03-1","17.0","17.1","17.3","17.4",
"17.5","18.0","18.1","18.2"]
```
Buggy versions can now be excluded when some modules or function/arity used. Limited to R16B03 and syntax_tools as first example. To be augmented, obviously.
May help for a smart package page.

**Comment by essen** [comment]
Thanks for the heads up. Sounds like it could be useful.

**Comment by essen** [comment]
FYI you can get the build status for Erlang.mk nightlies at https://builds.ninenines.eu/logs/erlang-dot-mk-packages/

The build is done using the local Erlang version (therefore subject to change) and takes at least 2 hours to run (without -j). The best we could do right now is testing against the last 3 major releases, so something like 19.1.5, 18.3.4.4 and 17.5.something. Otherwise it would take too long. I'm open to making that happen on the build servers. Then someone can take care of downloading the build results and building a page from them.

**Comment by crownedgrouse** [comment]
Thanks,
However my PR related to tags may help you to know if compile error is normal because project is not coded for some release or if it is an Erlang.mk issue.

**Comment by essen** [comment]
Yes I realize that. And I would love to have something take those logs (possibly with some extra info like Erlang version) and rule out expected failures. But I am not going to write it anytime soon. Contributions welcome. :-)

**Comment by essen** [comment]
To be clearer: regardless of the merge status of the tags PR, the test suite will probably not be the one cleaning up after itself for quite some time, in parts because it's not quite as easy as that to add.

**Comment by essen** [comment]
The way I see it, we can have the test suite run normally and produce these logs (possibly for the last 3 Erlang versions) every night. Then sometimes in the morning (like 9) a separate cron takes these logs and filters out expected build failures and formats them.

Note that while we currently only save the short log file, we can also save the entire failed test build; there's nothing problematic about that, and therefore run geas etc. on it.

**Comment by crownedgrouse** [comment]
Geas can use source file on projects, so we can know if a build is possible with an Erlang release and skip the build otherwise. So only erlang.mk real issue would be in logs. A log signaling the skip could be added too if needed. I can do a make geas_check target in geas plugin that exit 0 or 1 depending a variable specifying the target Erlang release.

**Comment by crownedgrouse** [comment]
So that the skip should be easier with a &&

**Comment by essen** [comment]
The problem is that the nightly takes up to 3 hours depending on the system. I would rather not it take longer. There aren't that many projects failing because of Erlang version either, and the ones that do fail because of one of their dependency, which means we would need to do something special to fetch dependencies first (we can do it, but the build is supposed to reproduce how people build applications).

**Comment by crownedgrouse** [comment]
I understand but geas is really fast and Mho it doesn't make sens to try to compile something we can know it can't, even if it is rare. If I understand there is now a fetch-deps target to download deps without compile it.
Workflow would be:
- Fetch deps
- Run geas on source && compile || log skip

**Comment by essen** [comment]
Fast is relative, when you got 20 dependencies to check.

Anyway the build simply does "make packages -k", so feel free to experiment and see if using geas speeds it up.

I don't think I have merged the fetch-deps PR yet, it will be merged in a few days though. So you can rely on it.

**Comment by crownedgrouse** [comment]
I thought I saw a PR for fetch-deps but it is maybe list-deps.

**Comment by crownedgrouse** [comment]
I will consider your last post, and experiment if I get some time left.

**Comment by essen** [comment]
Good luck. These kinds of experiments can be really time intensive. :-) It's also why I would much rather see some post-processing (quicker to setup and quicker to get useful data).

**Comment by essen** [comment]
Sorry I dropped the ball on this one, guys. Packages will be dropped eventually so most of what's here will not make it. Closing for now, thanks!
